### DISCUSSION TITLE
Cloning, COmmiting, and Pushing via SSH

### DISCUSSION THREAD

**Discussion by Mailstorm-ctrl** [jelmer/dulwich#1318] [Q&A]
I've searched and searched but I can't seem to figure this out. We have a github enterprise instance that requires SSO sign in. In order to do anything with git in a system context we must use SSH (There are apps, but they appear to be to complex for our use case). Reading seems to imply that I can use dulwich for this.

This is probably lack of experience but I just can't figure out how to do what I need to with SSH.

I've followed [this doc](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent) to get my SSH key authenticated. I tested ssh auth and it works, so I know it's not a key issue. Really the only thing I've tried is:

`dulwich.porcelain.clone("ssh://git@github.com:orgname/reponame.git", "C:\\temp\\repo")`

Which will make .git in the folder but then comes back with `dulwich.errors.HangupException: ssh: Could not resolve hostname ssh: No such host is known.` I know enough to know what this error means but I don't know what I need to do in Dulwich to correct it. THis post heavily implies that it "should just work" https://github.com/jelmer/dulwich/discussions/941

**Comment by Mailstorm-ctrl** [comment]
Seems like the uri github gives isn't correct. Replacing : with / works. Now I just need to figure out why I get errors about another process using a file in the repo.

**Reply by jelmer** [reply]
yep, that's right - it's either "ssh://git@github.com/orgname/reponame.git" (proper URL) /or/ "git@github.com:orgname/reponame.git" (rsync-style location)

**Reply by Mailstorm-ctrl** [reply]
Well ain't that something. If I could bother you about something else that'd be cool.

How can I specify a specific sshkey file to use when doing clones/commits/pushes? Ideally it's only used within the script and not a system wide change that would affect other scripts on the system.

**Reply by jelmer** [reply]
Dulwich just invokes SSH so that's something you would need to do on the SSH side. You can override the specific SSH command from Dulwich to e.g. pass in a key file.
